Oracle Exascale Feature - VM Cluster (#30568)
terraform-providers/terraform-provider-azurerm: website/docs/r/oracle_exascale_database_virtual_machine_cluster.html.markdown

---
subcategory: "Oracle"
layout: "azurerm"
page_title: "Azure Resource Manager: azurerm_oracle_exa_db_vm_cluster"
description: |-
  Manages a Exadata VM Cluster.
---

# azurerm_oracle_exa_db_vm_cluster

Manages a Exadata VM Cluster on Exascale Infrastructure.

## Example Usage

```hcl
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_oracle_exascale_db_storage_vault" "example" {
  name                = "example-exascale-db-storage-vault"
  display_name        = "example-exascale-db-storage-vault"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  description         = "description"

>   high_capacity_database_storage_input {

Reviewer: This was marked as resolved, but it isn't

  resource_group_name = azurerm_resource_group.example.name
}

resource "azurerm_subnet" "example" {
  name                 = "example-subnet"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefixes     = ["10.0.1.0/24"]

  delegation {
    name = "delegation"

    service_delegation {
      actions = [
        "Microsoft.Network/networkinterfaces/*",
        "Microsoft.Network/virtualNetworks/subnets/join/action",
      ]
      name = "Oracle.Database/networkAttachments"
    }
  }
}

resource "azurerm_oracle_exascale_database_virtual_machine_cluster" "example" {
  name                               = "example-exadb-vm-cluster"
  resource_group_name                = azurerm_resource_group.example.name
  location                           = azurerm_resource_group.example.location
  exascale_database_storage_vault_id = azurerm_oracle_exascale_database_storage_vault.example.id
  display_name                       = "example-exadb-vm-cluster"
  enabled_ecpu_count                 = 4
  hostname                           = "hostname"
  node_count                         = 2

>   shape                              = "ExaDbXS"

Reviewer 2: Should this be `EXADBXS`?

Author: The value is case-insensitive. We updated the documentation to use one style with a validation function.

  ssh_public_keys                    = [file("~/.ssh/id_rsa.pub")]
  subnet_id                          = azurerm_subnet.example.id
  total_ecpu_count                   = 10
  virtual_machine_file_system_storage {
    total_size_in_gb = 120
  }
  virtual_network_id = azurerm_virtual_network.example.id
}
```

## Arguments Reference

The following arguments are supported:

* `name` - (Required) The name which should be used for this Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created.

* `resource_group_name` - (Required) The name of the Resource Group where the Exadata VM Cluster should exist. Changing this forces a new Exadata VM Cluster to be created.

* `location` - (Required) The Azure Region where the Exadata VM Cluster should exist. Changing this forces a new Exadata VM Cluster to be created.

* `display_name` - (Required) The user-friendly name for the Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created. The name does not need to be unique.

> * `enabled_ecpu_count` - (Required) The number of ECPUs to enable for an Exadata VM cluster on Exascale Infrastructure. Changing this forces a new resource to be created.

Reviewer 3: ```suggestion
* `enabled_ecpu_count` - (Required) The number of ECPUs to enable for an Exadata VM cluster on Exascale Infrastructure. Possible values range between `8` and `200`, and must be divisible by `4`. Changing this forces a new resource to be created.
```

* `exascale_database_storage_vault_id` - (Required) The [OCID](https://docs.cloud.oracle.com/iaas/Content/General/Concepts/identifiers.htm) of the Exadata Database Storage Vault. Changing this forces a new Exadata VM Cluster to be created.

* `hostname` - (Required) The hostname for the Exadata VM Cluster on Exascale Infrastructure. Changing this forces a new Exadata VM Cluster to be created.

* `node_count` - (Required) The number of nodes in the Exadata VM cluster on Exascale Infrastructure. Changing this forces a new resource to be created.

> * `shape` - (Required) The shape of the Exadata VM cluster on Exascale Infrastructure resource.

Reviewer 2: Can you confirm if this can be updated?

* `ssh_public_keys` - (Required) The public key portion of one or more key pairs used for SSH access to the Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created.

* `subnet_id` - (Required) The ID of the subnet associated with the Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created.

* `total_ecpu_count` - (Required) The number of Total ECPUs for an Exadata VM cluster on Exascale Infrastructure.  Changing this forces a new resource to be created.

* `virtual_machine_file_system_storage` - (Required) A `virtual_machine_file_system_storage` block as defined below.

> * `virtual_network_id` - (Required) The ID of the Virtual Network associated with the Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created.

Reviewer 2: Could this be deduced from the `subnet_id`?

Reviewer 2: @evgenwk Could you please check this?

Author: The text of the documentation was updated.

Reviewer 2: In this case, since we can construct the parent virtual network id from a subnet id, I'd suggest to remove this from the schema.

> * `zones` - (Required) Cloud Exadata Infrastructure zones. Changing this forces a new Cloud Exadata Infrastructure to be created.

Reviewer 3: ```suggestion
* `zones` - (Required) Specifies a list of Availability Zones in which this Exadata VM Cluster should be located. Changing this forces a new Exadata VM Cluster to be created.
```

---

* `backup_subnet_cidr` - (Optional) The backup subnet CIDR of the Virtual Network associated with the Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created.

* `cluster_name` - (Optional) The cluster name for Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created.

* `data_collection_option` - (Optional) A `data_collection_option` block as defined below. Changing this forces a new Exadata VM Cluster to be created.

* `domain` - (Optional) The name of the OCI Private DNS Zone to be associated with the Exadata VM Cluster. This is required for specifying your own private domain name. Changing this forces a new Exadata VM Cluster to be created.

> * `grid_image_ocid` - (Required) Grid Setup will be done using this grid image ocid. Changing this forces a new resource to be created.

Reviewer 2: ```suggestion
* `grid_image_ocid` - (Required) The [OCID](https://docs.cloud.oracle.com/iaas/Content/General/Concepts/identifiers.htm) of the grid image that is used for grid setup. Changing this forces a new resource to be created.
```

* `license_model` - (Optional) The Oracle license model that applies to the Exadata VM Cluster, either `BringYourOwnLicense` or `LicenseIncluded`. Changing this forces a new Exadata VM Cluster to be created.

> * `network_security_group_cidrs` - (Optional) A `network_security_group_cidr` block as defined below. Changing this forces a new Exadata VM Cluster to be created.

Author: In another PR we have an opposite recommendation:
https://github.com/hashicorp/terraform-provider-azurerm/pull/30841#discussion_r2453967841

<img width="789" height="560" alt="20251028-150239-screenshot" src="https://github.com/user-attachments/assets/65e6036b-9b71-4ccf-b959-3a9e14f2b18d" />

> * `network_security_group_cidrs` - (Optional) A `network_security_group_cidr` block as defined below. Changing this forces a new Exadata VM Cluster to be created.

Reviewer 2: Confirmed, this should be singular.

* `private_zone_ocid` - (Optional) The The [OCID](https://docs.cloud.oracle.com/iaas/Content/General/Concepts/identifiers.htm) of the private zone in which you want DNS records to be created. Changing this forces a new Exadata VM Cluster to be created.

* `single_client_access_name_listener_port_tcp` - (Optional) The TCP Single Client Access Name (SCAN) port. Defaults to `1521`. Changing this forces a new Exadata VM Cluster to be created.

> * `single_client_access_name_listener_port_tcp_ssl` - (Optional) The TCPS Single Client Access Name (SCAN) port. Defaults to `2484`. Changing this forces a new Exadata VM Cluster to be created.

Reviewer 3: ```suggestion
* `single_client_access_name_listener_port_tcp_ssl` - (Optional) The TCPS Single Client Access Name (SCAN) port. Possible values range between `1024` and `8999`. Defaults to `2484`. Changing this forces a new Exadata VM Cluster to be created.
```

> * `system_version` - (Optional) Operating system version of the Exadata image. System version must be <= Db server major version (the first two parts of the DB server version eg 23.1.X.X.XXXX). Changing this forces a new resource to be created.

Reviewer: Additional info beyond a basic description and defaults, possible values, forcenew should generally go in a note.
```suggestion
* `system_version` - (Optional) Operating system version of the Exadata image. Changing this forces a new resource to be created.

~> **Note:** `system_version` must be less than or equal to the database server's major version (the first two parts of the database server version, e.g. 23.1.X.X.XXXX).
```

Author: Updated.

* `tags` - (Optional) A mapping of tags which should be assigned to the Exadata VM Cluster.

* `time_zone` - (Optional) The time zone of the Exadata VM Cluster. For details, see [Exadata Infrastructure Time Zones](https://docs.cloud.oracle.com/iaas/Content/Database/References/timezones.htm). Changing this forces a new Exadata VM Cluster to be created.

---

A `virtual_machine_file_system_storage` block supports the following:

* `total_size_in_gb` - (Required) Total Capacity.

---

A `data_collection_option` block supports the following:

* `diagnostics_events_enabled` - (Optional) Indicates whether diagnostic collection is enabled for the VM cluster/Cloud VM cluster/VMBM DBCS. Changing this forces a new Cloud VM Cluster to be created.

* `health_monitoring_enabled` - (Optional) Indicates whether health monitoring is enabled for the VM cluster / Cloud VM cluster / VMBM DBCS. Changing this forces a new Cloud VM Cluster to be created.

> * `incident_logs_enabled` - (Optional) Indicates whether incident logs and trace collection are enabled for the VM cluster / Cloud VM cluster / VMBM DBCS. Changing this forces a new Cloud VM Cluster to be created.

Reviewer 2: ```suggestion
* `incident_logs_enabled` - (Optional) Indicates whether incident logs and trace collection are enabled for the VM cluster, Cloud VM cluster or VMBM DBCS. Changing this forces a new Cloud VM Cluster to be created.
```

---

A `network_security_group_cidr` block exports the following:

* `destination_port_range` - (Required) A `destination_port_range` block as defined below.

* `source` - (Required) It is a range of IP addresses that a packet coming into the instance can come from. Changing this forces a new resource to be created.

---

A `destination_port_range` block exports the following:

* `max` - (Required) The maximum port number, which must not be less than the minimum port number. Changing this forces a new resource to be created.

* `min` - (Required) The minimum port number, which must not be greater than the maximum port number. Changing this forces a new resource to be created.

## Attributes Reference

In addition to the Arguments listed above - the following Attributes are exported: 

* `id` - The ID of the Exadata VM Cluster.

* `ocid` - The [OCID](https://docs.cloud.oracle.com/iaas/Content/General/Concepts/identifiers.htm) of the Exadata VM Cluster.

* `zone_ocid` - The [OCID](https://docs.oracle.com/en-us/iaas/Content/General/Concepts/identifiers.htm) of the zone the Exadata VM Cluster is associated with.

## Timeouts

> The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:

Reviewer 3: This needs to be updated to align with the latest [guidelines](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-new-resource.md#step-8-add-documentation-for-this-resource)
```suggestion
The `timeouts` block allows you to specify [timeouts](https://developer.hashicorp.com/terraform/language/resources/configure#define-operation-timeouts) for certain actions:
```

    }
  }
}

resource "azurerm_oracle_exascale_database_storage_vault" "example" {
  name                = "example-exascale-db-storage-vault"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  display_name        = "ExampleExascaleDbStorage"
  description         = "description"
  high_capacity_database_storage {
    total_size_in_gb = 300
  }
  additional_flash_cache_percentage = 100
  zones                             = ["3"]
}

resource "azurerm_oracle_exascale_database_virtual_machine_cluster" "example" {
  name                = "example-exadb-vm-cluster"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location

  exascale_database_storage_vault_id = azurerm_oracle_exascale_database_storage_vault.example.id
  display_name                       = "ExampleExascaleVmCluster"
  enabled_ecpu_count                 = 4
  hostname                           = "host"
  node_count                         = 2
  ssh_public_keys                    = [file("~/.ssh/id_rsa.pub")]
  subnet_id                          = azurerm_subnet.example.id

>   total_ecpu_count                   = 8

Reviewer 4: This example is incorrect. It has invalid `enabled_ecpu_count` -- the schema said minimum value is 8. It's also missing `grid_image_ocid`.

```suggestion
  enabled_ecpu_count                 = 8
  hostname                           = "host"
  grid_image_ocid                    = "ocid1.dbpatch.oc1.eu-amsterdam-1.anqw2ljrt5t4sqqad2xz3mnjtrszpuxrnkynaezs6elqzvu7t54iuq56tgkq"
  node_count                         = 2
  ssh_public_keys                    = [file("~/.ssh/id_rsa.pub")]
  subnet_id                          = azurerm_subnet.example.id
  total_ecpu_count                   = 16
```

Author: Updated

* `resource_group_name` - (Required) The name of the Resource Group where the Exadata VM Cluster should exist. Changing this forces a new Exadata VM Cluster to be created.

* `location` - (Required) The Azure Region where the Exadata VM Cluster should exist. Changing this forces a new Exadata VM Cluster to be created.

* `display_name` - (Required) The user-friendly name for the Exadata VM Cluster. Changing this forces a new Exadata VM Cluster to be created. The name does not need to be unique.

* `enabled_ecpu_count` - (Required) The number of ECPUs to enable for an Exadata VM cluster on Exascale Infrastructure. Possible values range between `8` and `200`, and must be divisible by `4`. Changing this forces a new resource to be created.

* `exascale_database_storage_vault_id` - (Required) The Azure Resource ID of the Exadata Database Storage Vault. Changing this forces a new Exadata VM Cluster to be created.

* `hostname` - (Required) The hostname for the Exadata VM Cluster on Exascale Infrastructure. Changing this forces a new Exadata VM Cluster to be created.

> * `grid_image_ocid` - (Required) The [OCID](https://docs.cloud.oracle.com/iaas/Content/General/Concepts/identifiers.htm) of the grid image that is used for grid setup. Changing this forces a new resource to be created. The Grid Image OCID value can be obtained using the [giMinorVersions API](https://learn.microsoft.com/en-us/rest/api/oracle/gi-minor-versions/list-by-parent?view=rest-oracle-2025-09-01&tabs=HTTP) and [giVersions API](https://learn.microsoft.com/en-us/rest/api/oracle/gi-versions/list-by-location?view=rest-oracle-2025-09-01&tabs=HTTP).

Reviewer 4: Use note block and elaborate how the APIs meant to be used. Also avoid `en-us` locale path when linking MS learn page.

```suggestion
* `grid_image_ocid` - (Required) The [OCID](https://docs.cloud.oracle.com/iaas/Content/General/Concepts/identifiers.htm) of the grid image that is used for grid setup. Changing this forces a new resource to be created.

-> **Note:** The Grid Image OCID value can be obtained by first listing the available `giVersion.name` using the [giVersions API](https://learn.microsoft.com/rest/api/oracle/gi-versions/list-by-location?view=rest-oracle-2025-09-01&tabs=HTTP): `az rest -m GET -u '/subscriptions/{subscriptionId}/providers/Oracle.Database/locations/{location}/giVersions?api-version=2025-09-01'` and then listing the available `gridImageOcid` for the `giVersionName`, `zone` and `shapeFamily` using the [giMinorVersions API](https://learn.microsoft.com/rest/api/oracle/gi-minor-versions/list-by-parent?view=rest-oracle-2025-09-01&tabs=HTTP): `az rest -m GET -u 'https://management.azure.com/subscriptions/{subscriptionId}/providers/Oracle.Database/locations/{location}/giVersions/s*Exadata.X11M+z*3+v*26.0.0.0/giMinorVersions?api-version=2025-09-01&shapeFamily=EXADB_XS&zone=1'````

Author: Updated
